src/middlewared/middlewared/plugins/smb_/groupmap.py

```diff
@@ -15,134 +20,334 @@ class Config:
+            elif g['sid'].startswith(localsid) and g['gid'] in range(544, 547):
+                rv['local_builtins'][gid] = g
+            elif g['sid'].startswith(localsid):
+                rv['local'][gid] = g
+            else:
+                rv['invalid'].append(g['sid'])
+
+        rv["localsid"] = localsid
+        return rv
+
+    @private
+    async def sync_builtins(self, groupmap):
+        idmap_backend = await self.middleware.call("smb.getparm", "idmap config *:backend", "GLOBAL")
+        idmap_range = await self.middleware.call("smb.getparm", "idmap config *:range", "GLOBAL")
+        payload = {"ADD": [], "MOD": [], "DEL": []}
+        must_reload = False
+
+        if idmap_backend != "tdb":
+            """
+            idmap_autorid and potentially other allocating idmap backends may be used for
+            the default domain. We do not want to touch how these are allocated.
+            """
+            return must_reload
+
+        low_range = int(idmap_range.split("-")[0].strip())
+        sid_lookup = {x["sid"]: x for x in groupmap.values()}
+
+        for b in SMBBuiltin:
+            sid = b.value[1]
+            rid = int(sid.split('-')[-1])
+            gid = low_range + (rid - 544)
+            entry = sid_lookup.get(sid, None)
+            if entry and entry['gid'] == gid:
+                # Value is correct, nothing to do.
+                continue
+
+            # If group type is incorrect, it entry must be deleted before re-adding.
+            elif entry and entry['gid'] != gid and entry['group_type_int'] != 4:
+                payload['DEL'].append({
+                    'sid': str(sid),
+                })
+                payload['ADD'].append({
+                    'sid': str(sid),
+                    'gid': gid,
+                    'group_type_str': 'local',
+                    'nt_name': b.value[0][8:].capitalize()
+                })
+            elif entry and entry['gid'] != gid:
+                payload['MOD'].append({
+                    'sid': str(sid),
+                    'gid': gid,
+                    'group_type_str': 'local',
+                    'nt_name': b.value[0][8:].capitalize()
+                })
+            else:
+                payload['ADD'].append({
+                    'sid': str(sid),
+                    'gid': gid,
+                    'group_type_str': 'local',
+                    'nt_name': b.value[0][8:].capitalize()
+                })
+
+        await self.batch_groupmap(payload)
+        if (await self.middleware.call('smb.validate_groupmap_hwm', low_range)):
+            must_reload = True
+
+        return must_reload
+
+    @private
+    async def batch_groupmap(self, data):
+        for op in ["ADD", "MOD", "DEL"]:
+            if data.get(op) is not None and len(data[op]) == 0:
+                data.pop(op)
+
+        payload = json.dumps(data)
+        out = await run([SMBCmd.NET.value, '--json', 'groupmap', 'batch_json', f'data={payload}'], check=False)
```

Comment:
I'm assuming that if you have a situation where there is a `DEL` and an `ADD` operation, the underlying `batch_json` operation will process those in order? Or does order not matter?

Reply:
Right. We iterate through a lookup table with function pointers:
```
const struct {
	const char *name;
	int (*fn)(struct net_context *c, int argc, const char **argv);
} optable[] = {
	{ "DEL", net_groupmap_delete },
	{ "ADD", net_groupmap_add },
	{ "MOD", net_groupmap_modify },
};
```

Then use `name` to attempt to retrieve a json array of data to iterate through and pass to `fn`. So we get "DEL", then "ADD", then "MOD".